daemons/fenced/pacemaker-fenced.c

```diff
@@ -765,6 +765,10 @@ update_cib_stonith_devices_v2(const char *event, xmlNode * msg)
     }
 
     if(needs_update) {
+        if (g_hash_table_size(device_list)) {
+            free_device_list();
+            init_device_list();
+        }
```

Comment:
How is this? https://github.com/inouekazu/pacemaker/commit/4fa63fc921e4c1235cdb23cf55f557481b2891a8

Reply:
Not there -- that can be called via other paths that shouldn't remove a device if it's registered at all.

I was thinking instead of free_device_list()+init_device_list(), do g_hash_table_foreach_remove().

Reply:
I also considered that specification (https://github.com/inouekazu/pacemaker/commit/74d8d552ef8332ac20fd44218c97d92a5ecf5ed5), but api_registered is true when needs_update is true, so fence1-vrish will not be deleted...
```
# pcs cluster cib-push cib-2.xml --config

# tail -F /var/log/pacemaker/pacemaker.log | grep -e stonith_device_register -e stonith_device_remove -e update_cib_stonith_devices_v2 -e debug_print_device -e is_api_unregistered
Jan 14 15:24:52 node-2 pacemaker-fenced    [145342] (update_cib_stonith_devices_v2@pacemaker-fenced.c:790)   debug: needs_update[TRUE]
Jan 14 15:24:52 node-2 pacemaker-fenced    [145342] (debug_print_device@pacemaker-fenced.c:719)              debug: device_list: id[fence1-virsh] api_registered[TRUE]
Jan 14 15:24:52 node-2 pacemaker-fenced    [145342] (is_api_unregistered@pacemaker-fenced.c:731)             debug: not removing[fence1-virsh] because api_registered is TRUE
Jan 14 15:24:52 node-2 pacemaker-fenced    [145342] (update_cib_stonith_devices_v2@pacemaker-fenced.c:796)   info: Updating device list from CIB: (meta) attribute deleted from resource
Jan 14 15:24:52 node-2 pacemaker-fenced    [145342] (stonith_device_register@fenced_commands.c:1197)         notice: Added 'fence1-ipmilan' to device list (2 active devices)

Jan 14 15:24:52 node-2 pacemaker-fenced    [145342] (update_cib_stonith_devices_v2@pacemaker-fenced.c:790)   debug: needs_update[FALSE]
Jan 14 15:24:52 node-2 pacemaker-fenced    [145342] (debug_print_device@pacemaker-fenced.c:719)              debug: device_list: id[fence1-virsh] api_registered[TRUE]
Jan 14 15:24:52 node-2 pacemaker-fenced    [145342] (debug_print_device@pacemaker-fenced.c:719)              debug: device_list: id[fence1-ipmilan] api_registered[FALSE]

Jan 14 15:24:52 node-2 pacemaker-fenced    [145342] (update_cib_stonith_devices_v2@pacemaker-fenced.c:790)   debug: needs_update[FALSE]
Jan 14 15:24:52 node-2 pacemaker-fenced    [145342] (debug_print_device@pacemaker-fenced.c:719)              debug: device_list: id[fence1-virsh] api_registered[TRUE]
Jan 14 15:24:52 node-2 pacemaker-fenced    [145342] (debug_print_device@pacemaker-fenced.c:719)              debug: device_list: id[fence1-ipmilan] api_registered[FALSE]
 :
Jan 14 15:24:53 node-2 pacemaker-fenced    [145342] (update_cib_stonith_devices_v2@pacemaker-fenced.c:790)   debug: needs_update[FALSE]
Jan 14 15:24:53 node-2 pacemaker-fenced    [145342] (debug_print_device@pacemaker-fenced.c:719)              debug: device_list: id[fence1-virsh] api_registered[TRUE]
Jan 14 15:24:53 node-2 pacemaker-fenced    [145342] (debug_print_device@pacemaker-fenced.c:719)              debug: device_list: id[fence1-ipmilan] api_registered[FALSE]
Jan 14 15:24:53 node-2 pacemaker-fenced    [145342] (stonith_device_register@fenced_commands.c:1173)         debug: Device 'fence1-ipmilan' already in device list (2 active devices)
 :
Jan 14 15:24:54 node-2 pacemaker-fenced    [145342] (update_cib_stonith_devices_v2@pacemaker-fenced.c:790)   debug: needs_update[FALSE]
Jan 14 15:24:54 node-2 pacemaker-fenced    [145342] (debug_print_device@pacemaker-fenced.c:719)              debug: device_list: id[fence1-virsh] api_registered[TRUE]
Jan 14 15:24:54 node-2 pacemaker-fenced    [145342] (debug_print_device@pacemaker-fenced.c:719)              debug: device_list: id[fence1-ipmilan] api_registered[TRUE]
Jan 14 15:24:54 node-2 pacemaker-fenced    [145342] (stonith_device_remove@fenced_commands.c:1242)           trace: Not removing 'fence1-virsh' from device list (2 active) because still registered via: cib

Jan 14 15:24:54 node-2 pacemaker-fenced    [145342] (update_cib_stonith_devices_v2@pacemaker-fenced.c:790)   debug: needs_update[FALSE]
Jan 14 15:24:54 node-2 pacemaker-fenced    [145342] (debug_print_device@pacemaker-fenced.c:719)              debug: device_list: id[fence1-virsh] api_registered[FALSE]
Jan 14 15:24:54 node-2 pacemaker-fenced    [145342] (debug_print_device@pacemaker-fenced.c:719)              debug: device_list: id[fence1-ipmilan] api_registered[TRUE]
 :
Jan 14 15:24:54 node-2 pacemaker-fenced    [145342] (update_cib_stonith_devices_v2@pacemaker-fenced.c:790)   debug: needs_update[FALSE]
Jan 14 15:24:54 node-2 pacemaker-fenced    [145342] (debug_print_device@pacemaker-fenced.c:719)              debug: device_list: id[fence1-virsh] api_registered[FALSE]
Jan 14 15:24:54 node-2 pacemaker-fenced    [145342] (debug_print_device@pacemaker-fenced.c:719)              debug: device_list: id[fence1-ipmilan] api_registered[TRUE]
```

Reply:
Hmm, that's disappointing. The executor will register the device via the API when starting it in execd_stonith_start(), in case the fencer hasn't processed the configuration yet, so that's probably how it got API-registered.

I just noticed one other case where a device is API-registered only: the implicit device created for a watchdog (near the end of main()). We can't remove that one.

I'm thinking we could take another approach: before updating, go through all existing devices in the table, and if they're CIB-registered, mark them "dirty" (via a new boolean in stonith_device_t). stonith_device_register() (which cib_device_update() calls when a CIB device should be used) could always clear that boolean. Then after updating, go through the table again, and remove any devices still marked dirty.

That would avoid deleting anything that was API-registered only, whether something external, the watchdog, or the executor.

We could remove the TODO in cib_device_update() that's related to this problem. That function could also just return instead of goto update_done (which wouldn't be needed anymore), since the device would get removed later for being dirty.

I'm thinking we should do it in cib_devices_update() rather than update_cib_stonith_devices_v2(), since the other calls would have the same problem.

Reply:
I'm going to close this PR because I requested a fix that reflects your comment (https://github.com/ClusterLabs/pacemaker/pull/2285), but I do have one comment.
https://github.com/ClusterLabs/pacemaker/blob/9b131e309d2fdb466b154d8d9e98ab0d1a6e43c7/daemons/fenced/pacemaker-fenced.c#L1508-L1539
The stonith_watchdog_timeout_ms is updated only when a "CIB change" is received (setup_cib() -> update_cib_cache_cb()), so it remains "0" at the timing of L1520. I'll consider it.

Reply:
That's interesting ... I wonder how it ever worked. @wenningerk , what do you think?

Reply:
That is working via setup_cib cmds->query and cmds->register_callback.
Somehow at least. Iirc this is the part that is racy and can lead to neverending loops of cib-queries.

Reply:
I see update_cib_cache_cb() uses cib_sync_call when local_cib==NULL, so maybe it somehow gets called before that point. I'm not sure though, and we may not want to rely on that.

Reply:
It is ugly but it gets called every time. The issue is rather that the updating calls something that does a dispatch and if there is a real cib-update (diff) in between that iirc is dropping - leading to a subsequent diff not matching the base version -  leading to a full update request - and that could happen in an endless loop iirc.

Reply:
I had that fixed somehow on my watchdog-fencing-for-certain-devices branch. But I don't remember if it was less ugly. Was actually not the intention of the branch to fix that. The additional communication with with tengine I think just made the hang so likely that I had to do something.

Reply:
I requested a fix for this bug. (#2292)